ruma-common/src/push.rs

```diff
@@ -103,6 +106,106 @@ pub struct PatternedPushRule {
     pub pattern: String,
 }
 
+/// Like `PushRule`, but may represent any kind of push rule
+/// thanks to `pattern` and `conditions` being optional.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct AnyPushRule {
+    /// The actions to perform when this rule is matched.
+    pub actions: Vec<Action>,
+
+    /// Whether this is a default rule, or has been set explicitly.
+    pub default: bool,
+
+    /// Whether the push rule is enabled or not.
+    pub enabled: bool,
+
+    /// The ID of this rule.
+    pub rule_id: String,
+
+    /// The conditions that must hold true for an event in order for a rule to be applied to an event. A rule with no conditions always matches.
+    /// Only applicable to underride and override rules.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub conditions: Option<Vec<PushCondition>>,
+
+    /// The glob-style pattern to match against. Only applicable to content rules.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub pattern: Option<String>,
+}
+
+impl From<PushRule> for AnyPushRule {
+    fn from(push_rule: PushRule) -> Self {
+        let PushRule { actions, default, enabled, rule_id } = push_rule;
+        AnyPushRule { actions, default, enabled, rule_id, pattern: None, conditions: None }
+    }
+}
+
+impl From<PatternedPushRule> for AnyPushRule {
+    fn from(push_rule: PatternedPushRule) -> Self {
+        let PatternedPushRule { actions, default, enabled, rule_id, pattern } = push_rule;
+        AnyPushRule { actions, default, enabled, rule_id, pattern: Some(pattern), conditions: None }
+    }
+}
+
+impl From<ConditionalPushRule> for AnyPushRule {
+    fn from(push_rule: ConditionalPushRule) -> Self {
+        let ConditionalPushRule { actions, default, enabled, rule_id, conditions } = push_rule;
+        AnyPushRule {
+            actions,
+            default,
+            enabled,
+            rule_id,
+            pattern: None,
+            conditions: Some(conditions),
+        }
+    }
+}
+
+impl From<AnyPushRule> for PushRule {
+    fn from(push_rule: AnyPushRule) -> Self {
+        let AnyPushRule { actions, default, enabled, rule_id, .. } = push_rule;
+        PushRule { actions, default, enabled, rule_id }
+    }
+}
+
+/// An error that happens when `AnyPushRule` cannot be converted
+/// into a more specific push rule type.
+#[derive(Debug)]
+pub struct AnyPushRuleConversionError;
```

Comment:
Could you make this an enum of `MissingConditions | MissingPattern` and implement `Display` & `Error` for it?

Reply:
Or make it two unit structs, whatever you think makes more sense.

Reply:
Oh, yeah, I was going to ask how you usually deal with errors, since there are so many options, but forgot!

Reply:
Also maybe it should be two separate errors? `try_from` for `PatternedPushRule` cannot possibly produce an error that says that a condition is missing.

Reply:
> Oh, yeah, I was going to ask how you usually deal with errors, since there are so many options, but forgot!

It depends. It the are used for strictly independent things, I like having separate error types too. If there is higher level functionality built on top of that, it often becomes too cumbersome to have a separate error types for every function with a different set of possible error conditions.

> Also maybe it should be two separate errors?

That's what I meant by "two unit structs". Let's go with this option.

Reply:
> That's what I meant by "two unit structs". Let's go with this option.

Oops, sorry! I blame 3am 😅